### PR TITLE
Wip/lapi `cscli capi` : allow interraction with central API

### PR DESCRIPTION
**Pull request by buixor**
- `cscli capi register`
 - `cscli capi status`
 - make pkg/apiclient have a configurable prefix

### CODE REVIEW

File: cmd/crowdsec-cli/capi.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/davecgh/go-spew/spew"
+	"github.com/go-openapi/strfmt"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+
+	"gopkg.in/yaml.v2"
+)
+
+var APIBaseURL string = "https://api.dev.crowdsec.net/"
```
Comment (AlteredCoder) [inline cmd/crowdsec-cli/capi.go:23]: don't we want this to be part of the configuration?

File: cmd/crowdsec-cli/capi.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+var APIURLPrefix string = "v2"
```
Comment (AlteredCoder) [inline cmd/crowdsec-cli/capi.go:24]: don't we want this to be part of the configuration?

File: cmd/crowdsec-cli/capi.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+			Client = apiclient.NewClient(nil)
+
+			id, err := generateID()
+			if err != nil {
+				log.Fatalf("unable to generate machine id: %s", err)
+			}
+			log.Printf("your machine ID : %s", id)
+			password := strfmt.Password(generatePassword(passwordLength))
+			log.Printf("your password : %s", password)
+			_, err = Client.Auth.RegisterWatcher(context.Background(), models.WatcherRegistrationRequest{
```
Comment (AlteredCoder) [inline cmd/crowdsec-cli/capi.go:57]: do we want to check that `resp.StatusCode == 200` ?

File: cmd/crowdsec-cli/capi.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+				Login:    id,
+				Password: password.String(),
+				URL:      APIBaseURL,
+			}
+			apiConfigDump, err := yaml.Marshal(apiCfg)
+			if err != nil {
+				log.Fatalf("unable to marshal api credentials: %s", err)
+			}
+			if dumpFile != "" {
+				err = ioutil.WriteFile(dumpFile, apiConfigDump, 0644)
```
Comment (AlteredCoder) [inline cmd/crowdsec-cli/capi.go:86]: can you use `0600` please ?
Also, can you set the same rights in `machines.go` please ?

File: cmd/crowdsec-cli/capi.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+				log.Fatalf("failed to get scenarios : %s", err.Error())
+			}
+			password := strfmt.Password(csConfig.API.Server.OnlineClient.Credentials.Password)
+			t := models.WatcherAuthRequest{
+				MachineID: &csConfig.API.Server.OnlineClient.Credentials.Login,
+				Password:  &password,
+				Scenarios: scenarios,
+			}
+			Client = apiclient.NewClient(nil)
+			resp, err := Client.Auth.AuthenticateWatcher(context.Background(), t)
```
Comment (AlteredCoder) [inline cmd/crowdsec-cli/capi.go:132]: do we want to check that `resp.StatusCode == 200` ?

File: pkg/apiclient/auth_service.go [hunk 6]
```diff
@@ -27,7 +29,9 @@ func (s *AuthService) UnregisterWatcher(ctx context.Context) (*Response, error)
 
 func (s *AuthService) RegisterWatcher(ctx context.Context, registration models.WatcherRegistrationRequest) (*Response, error) {
 
-	u := "v1/watchers"
+	u := fmt.Sprintf("%s/watchers", s.client.URLPrefix)
+	log.Printf("->> %s", u)
```
Comment (AlteredCoder) [inline pkg/apiclient/auth_service.go:33]: To remove ?

File: pkg/apiclient/client.go [hunk 7]
```diff
@@ -22,13 +23,13 @@ type ApiClient struct {
...
 	/*config stuff*/
 	BaseURL   *url.URL
+	URLPrefix string
 	UserAgent string
 	/*exposed Services*/
 	Decisions *DecisionsService
 	Alerts    *AlertsService
 	Auth      *AuthService
 	Metrics   *MetricsService
-	// Consensus *ApiConsensus
```
Comment (AlteredCoder) [inline pkg/apiclient/client.go:31]: To remove ?

File: pkg/apiclient/client_http.go [hunk 8]
```diff
@@ -56,7 +58,16 @@ func (c *ApiClient) Do(ctx context.Context, req *http.Request, v interface{}) (*
 	if c.UserAgent != "" {
 		req.Header.Add("User-Agent", c.UserAgent)
 	}
+	if log.GetLevel() >= log.TraceLevel {
```
Comment (AlteredCoder) [inline pkg/apiclient/client_http.go:61]: is this dump redundant with the same code in `refreshJwtToken()` used by JWTtransport ? if yes, maybe we can remove it from the `refreshJwtToken()` ?

File: pkg/apiclient/client_http.go [hunk 9]
```diff
@@ -56,7 +58,16 @@ func (c *ApiClient) Do(ctx context.Context, req *http.Request, v interface{}) (*
...
+		dump, _ := httputil.DumpRequest(req, true)
+		log.Tracef("request: %s", string(dump))
+	}
+
 	resp, err := c.client.Do(req)
+	if log.GetLevel() >= log.TraceLevel {
```
Comment (AlteredCoder) [inline pkg/apiclient/client_http.go:67]: is this dump redundant with the same code in `refreshJwtToken()` used by JWTtransport ? if yes, maybe we can remove it from the `refreshJwtToken()` ?

File: cmd/crowdsec-cli/capi.go (hunk 3 above)
Comment (buixor) [reply cmd/crowdsec-cli/capi.go:58]: It is already done by the `Do()` method

File: cmd/crowdsec-cli/capi.go (hunk 4 above)
Comment (buixor) [reply cmd/crowdsec-cli/capi.go:87]: yes

File: cmd/crowdsec-cli/capi.go (hunk 5 above)
Comment (buixor) [reply cmd/crowdsec-cli/capi.go:133]: done in `Do()`

**Review by buixor — COMMENTED** [review]
lgtm, wait for cross review

**Pull request merged**